stan/math/prim/functor/apply.hpp

```diff
@@ -0,0 +1,50 @@
+#ifndef STAN_MATH_PRIM_SCAL_FUNCTOR_APPLY_HPP
+#define STAN_MATH_PRIM_SCAL_FUNCTOR_APPLY_HPP
+
+#include <functional>
+#include <tuple>
+#include <utility>
+
+namespace stan {
+namespace math {
+/*
+ * Invoke the functor f with arguments given in t and indexed in the index
+ * sequence I
+ *
+ * @tparam F Type of functor
+ * @tparam Tuple Type of tuple containing arguments
+ * @tparam I Index sequence going from 0 to std::tuple_size<T>::value - 1
+ * inclusive
+ * @param f functor callable
+ * @param t tuple of arguments
+ * @param i placeholder variable for index sequence
+ */
+template <class F, class Tuple, std::size_t... I>
+constexpr decltype(auto) apply_impl(F&& f, Tuple&& t,
+                                    std::index_sequence<I...> i) {
+  return f(std::forward<decltype(std::get<I>(t))>(std::get<I>(t))...);
+}
```

Comment:
Should this be internal?

Reply:
Hmm, I don't think so. It's meant to be used by any function in math (currently used by adj_jac_apply and here).

It's not intended to be exposed to the Stan language.

It'll get replaced by std::apply as soon as we get the next C++ stuff. I'll leave this decision to you. Certainly easy enough to just add an internal:: if we want.